File: jscomp/test/test_primitive.js
'use strict';

var Curry = require("../../lib/js/curry.js");
var Caml_array = require("../../lib/js/caml_array.js");
var Pervasives = require("../../lib/js/pervasives.js");
var CamlinternalLazy = require("../../lib/js/camlinternalLazy.js");

function a4(prim) {
  return [
          "File \"test_primitive.ml\", line 30, characters 9-19",
          prim
        ];
}

function a5(prim) {
  return [
          31,
          prim
        ];
}

function a6(prim) {
  return [
          /* tuple */[
            "test_primitive.ml",
            32,
            9,
            19
          ],
          prim
        ];
}

var test_float = 3;

var test_abs = Math.abs(3.0);

var v = /* array */[
  1.0,
  2.0
];

var xxx = "a";

var a = /* "a" */97;

function u(b) {
  if (b) {
    Pervasives.print_int(1);
    return 32;
  } else {
    return 7;
  }
}

function f2(h, b, _) {
  return Curry._1(h, b ? 32 : 7);
}

Caml_array.caml_array_set(v, 1, 3.0);

var unboxed_x = /* record */[
  /* u */0,
  /* v */0
];

function gg(x) {
  x[/* u */0] = 0;
  return /* () */0;
}

function f(x) {
  return x.length;
}

function is_lazy_force(x) {
  var tag = x.tag | 0;
  if (tag === 250) {
    return x[0];
  } else if (tag === 246) {
    return CamlinternalLazy.force_lazy_block(x);
  } else {
    return x;
  }
}

function fib(n) {
  if (n === 0 || n === 1) {
    return 1;
  } else {
    var fib1 = fib(n - 1 | 0);
    var fib2 = fib(n - 2 | 0);
    return (fib1 + fib2 | 0) + 3 | 0;
  }
}

var a0 = "File \"test_primitive.ml\", line 26, characters 9-16";

var a1 = "test_primitive.ml";

var a2 = 28;

var a3 = "Test_primitive";

var xx = /* tuple */[
  0,
  0
];

exports.a0 = a0;
exports.a1 = a1;
exports.a2 = a2;
exports.a3 = a3;
exports.a4 = a4;
exports.a5 = a5;
exports.a6 = a6;
exports.test_float = test_float;
exports.test_abs = test_abs;
exports.v = v;
exports.xxx = xxx;
exports.a = a;
exports.u = u;
exports.f2 = f2;
exports.xx = xx;
exports.unboxed_x = unboxed_x;
exports.gg = gg;
exports.f = f;
exports.is_lazy_force = is_lazy_force;
exports.fib = fib;
/*  Not a pure module */

File: jscomp/test/test_primitive.ml
(* include ( module X : sig val fib : int -> int end = struct let rec fib =
   function *)
(* | 0 | 1 -> 1 *)
(* | n -> fib (n - 1) + fib (n - 2) *)

(* let u = List.length *)
(* end ) *)

(* type u = Leaf | Node of u * u *)

(* let rec sum = function *)
(* | Leaf -> 1 *)
(* | Node (l,r) -> sum l + sum r *)

(* let leaf = Leaf *)
(* let node l r = Node (l,r) *)

(* let f = Math.abs 3.0 Original definition at externs.zip//es3.js:888 var
   Math=require("Math"); *)
(* import List from "List" *)
(* let g f x = x |> f *)

(* raise Not_found;; *)
(* Test alone next time*)
let a0 = __LOC__
let a1 = __MODULE__
let a2 = __LINE__
let a3 = __MODULE__
let a4 = __LOC_OF__
let a5 = __LINE_OF__
let a6 = __POS_OF__
let test_float = float 3
let test_abs = abs_float 3.0

(* Clflags.dump_lambda:=true *)
let v = [|1.0; 2.0|]
let xxx = "a"
let a = xxx.[0]

let u b =
  let a = if b then (print_int 1 ; 32) else 3 + 4 in
  a

let f h b () = h (if b then (print_int 1 ; 32) else 3 + 4)
let f2 h b () = h (if b then 32 else 3 + 4)

[@@@warning "-26"]

let f2 h b () =
  h
    ( if b then
      let c = 3 in
      32
    else 3 + 4 )

let () = v.(1) <- 3.0

type u = {mutable u: float; v: float}

let xx = (0., 0.)
let unboxed_x = {u= 0.; v= 0.}
let f {u} = u
let gg x = x.u <- 0.
let f (x : string) = String.length x
let is_lazy_force x = Lazy.force x

let rec fib (n : int) =
  match n with
  | 0 | 1 -> 1
  | n ->
      let fib1 = fib (n - 1) in
      let fib2 = fib (n - 2) in
      let fib3 = 3 in
      fib1 + fib2 + fib3
